src/main/java/com/relayrides/pushy/apns/ApnsClient.java

```diff
@@ -0,0 +1,749 @@
+                                ApnsClient.this.connectionReadyPromise.tryFailure(future.cause());
+                            }
+                        }
+                    });
+
+                    connectFuture.channel().closeFuture().addListener(new GenericFutureListener<ChannelFuture> () {
+
+                        @Override
+                        public void operationComplete(final ChannelFuture future) throws Exception {
+                            // We always want to try to fail the "connection ready" promise if the connection closes; if
+                            // it has already succeeded, this will have no effect.
+                            ApnsClient.this.connectionReadyPromise.tryFailure(
+                                    new IllegalStateException("Channel closed before HTTP/2 preface completed."));
+
+                            synchronized (ApnsClient.this.bootstrap) {
+                                ApnsClient.this.connectionReadyPromise = null;
+
+                                if (ApnsClient.this.reconnectionPromise != null) {
+                                    log.debug("Disconnected. Next automatic reconnection attempt in {} seconds.", ApnsClient.this.reconnectDelay);
+
+                                    future.channel().eventLoop().schedule(new Runnable() {
+
+                                        @Override
+                                        public void run() {
+                                            log.debug("Attempting to reconnect.");
+                                            ApnsClient.this.connect(host, port);
+                                        }
+                                    }, ApnsClient.this.reconnectDelay, TimeUnit.SECONDS);
+
+                                    ApnsClient.this.reconnectDelay = Math.min(ApnsClient.this.reconnectDelay, MAX_RECONNECT_DELAY);
+                                }
+                            }
+                        }
+                    });
+
+                    this.connectionReadyPromise.addListener(new GenericFutureListener<ChannelFuture>() {
+
+                        @Override
+                        public void operationComplete(final ChannelFuture future) throws Exception {
+                            if (future.isSuccess()) {
+                                synchronized (ApnsClient.this.bootstrap) {
+                                    if (ApnsClient.this.reconnectionPromise != null) {
+                                        log.info("Connection to {} restored.", future.channel().remoteAddress());
+                                        ApnsClient.this.reconnectionPromise.trySuccess();
+                                    } else {
+                                        log.info("Connected to {}.", future.channel().remoteAddress());
+                                    }
+
+                                    ApnsClient.this.reconnectDelay = INITIAL_RECONNECT_DELAY;
```

Comment:
And another useful feature we've found is to switch to a different max reconnectDelay (and max retries, see above) once the initial connect is successfully established -- ie. we have now determined that the SSL cert works ok.